File: Scripts/import_macros.py
import os
from re import sub
import subprocess

macro_dirs = "./Scripts/macros/"


def getMacoFilse(dir):

    ret = list()

    for root, dir, files in os.walk(dir):
        # print(root, dir, files, "\n")
        for file in files:
            if file.endswith(".lua"):
                name = file.removesuffix(".lua")
                path = root + file
                ret.append((name, path))

    return ret


if __name__ == '__main__':
    files = getMacoFilse(macro_dirs)
    # print(files)

    ts = list()

    for name, path in files:
        print(name, path)

        precmd = ["xmake", "m"]
        macro_path_arg = "--import="+path
        precmd.append(macro_path_arg)
        precmd.append(name)

        # cmd = "xmake m --import="+path+" " + name

        subprocess.run(precmd)

File: Scripts/import_macros.lua
require "lfs"
require "os"

local macros_dir = "./Scripts/macros"

local function getpathes(rootpath, pathes)
    pathes = pathes or {}
    for entry in lfs.dir(rootpath) do
        -- print(entry)
        -- ~= !=
        if entry ~= '.' and entry ~= '..' then
            local path = rootpath .. '/' .. entry
            local attr = lfs.attributes(path)
            assert(type(attr) == 'table')

            -- if attr.mode == 'directory' then
            --     getpathes(path, pathes)
            -- else
            pathes[entry] = path
        end
    end
    return pathes
end

local function getPreWords(file)
    local ret = ""
    for i = 1, #file do
        -- print(i)
        -- print("sub," .. i .. "," .. i)
        -- sub(s ,i, i) get the string [i]
        local k = string.sub(file, i, i)
        -- print(k)

        -- print(k .. "--" .. file)

        if k == '.' then
            break
        end

        ret = ret .. k
    end
    return ret
end

local function main(argv)
    -- files = getpathes(argv[1])
    local files = getpathes(macros_dir)

    for n, f in pairs(files) do
        print(n .. ":" .. f)
        n = getPreWords(n)

        local cmd = "xmake m --import=" .. f .. " " .. n
        os.execute(cmd)
    end
end

main(arg)
